src/sources/ascii.jl

```diff
@@ -0,0 +1,190 @@
+        metadata=xy_metadata,
+        crs=crs,
+        mappedcrs=mappedcrs,
+        dim=Y()
+    )
+
+    x = X(xlookup)
+    y = Y(ylookup)
+
+    return x,y
+
+end
+
+missingval(ap::ASCIIparams) = params(ap)[:nodatavalue]
+
+# no metadata for now
+DD.metadata(ap::ASCIIparams) = Metadata{ASCIIfile}()
+DD.ndims(ap::ASCIIparams) = 2
+
+# Base util methods
+
+Base.eltype(p::ASCIIparams{T}) where T = T
+
+function Base.size(ap::ASCIIparams)
+    params(ap)[:ncols], params(ap)[:nrows]
+end
+
+Base.Array(ap::ASCIIparams) = _asciigrid(a -> Array(a), ap)
+
+# Array
+#######
+
+function FileArray(ap::ASCIIparams, filename = filename(ap); kw...)
+    size_ = size(ap)
+    eachchunk = DiskArrays.GridChunks(size_, size_)
+    haschunks = DiskArrays.Unchunked()
+    T = eltype(ap)
+    N = ndims(ap)
+    FileArray{ASCIIfile, T, N}(filename, size_; eachchunk, haschunks, kw...)
+end
+
+# Base i/o methods
+###################
+
+# data (ASCIIfile) and metadata (ASCIIparams) objects are separate 
+# so data is opened using _asciigrid
+function Base.open(f::Function, A::FileArray{ASCIIfile}, key...; write = A.write)
+    _asciigrid(dat -> f(RasterDiskArray{ASCIIfile}(dat, A.eachchunk, A.haschunks)), A; write)
+end
+
+# also called by _open(f, fa::FileArray{ASCIIfile})
+function _open(f, ::Type{ASCIIfile}, filename; write = false, kw...)
+    isfile(filename) || _filenotfound_error(filename)
+    _open(f, ASCIIfile, ASCIIparams(filename; write))
+end
+
+_open(f, ::Type{ASCIIfile}, ap::ASCIIparams; kw...) = f(ap)
+
+function Base.write(filename::String, ::Type{ASCIIfile}, A::AbstractRaster{T, 2}) where T
+    # handle nodatavalue:
+    # set to default if missingval is not defined
+    if ismissing(missingval(A)) || isnothing(missingval(A))
+        A = replace_missing(A, -9999.0)
+    end
+
+    correctedA = permutedims(A, (X, Y)) |>
+        a -> reorder(a, (X(ForwardOrdered()), Y(ReverseOrdered())))
+
+    #build header
+    ncols, nrows = size(A)
+
+    xbounds, ybounds = bounds(A, (X, Y))
+    
+    xll = min(xbounds...)
+    yll = min(ybounds...)
+
+    dx = abs((xbounds[1] - xbounds[2])/ncols)
+    dy = abs((ybounds[1] - ybounds[2])/nrows)
+
+    type = eltype(A)
+
+    nodatavalue = type(missingval(A))
+
+    pars = (
+        ncols = ncols,
+        nrows = nrows,
+        xll = xll,
+        yll = yll,
+        dx = dx,
+        dy = dy,
+        nodatavalue = nodatavalue
+    )
+
+    ASCIIrasters.write_ascii(filename, _flip(correctedA, (nrows, ncols), type), pars)
+
+    return filename
+end
+
+# Utilities
+###########
+
+function _asciigrid(f::Function, ap::Union{ASCIIparams, FileArray}; kw...)
+    _asciigrid(f, filename(ap), eltype(ap), size(ap); kw...)
+end
+
+function _asciigrid(f, filename::AbstractString, T::Type, size::Tuple; write = false, kw...)
+    dat, pars = open(filename, "r") do io
+        dat, pars = ASCIIrasters.read_ascii(filename; lazy = false)
+        # dat is a nr x nc matrix, we want a nc x nr matrix for use
```

Comment:
Why flip this? We can just change the dimension order. 

It's good to minimise allocations in case the files are very large.

Reply:
Indeed, thanks !